### PR TITLE
Add: User Preferences

### PR DESCRIPTION
**Pull request by martinmckenna**
## Description

Removes local storage for theme and spacing and instead observes user preferences

## Type of Change
- Non breaking change ('update', 'change')

## Applicable E2E Tests

N/A

## Note to Reviewers

1. The high file-change count is mostly due to tests having to be updated.

## Highlights and Updates

1. New `<PreferenceToggle />` component, which follows a children-as-a-function react pattern. The current preference and callback to toggle the preference are passed down to the children.
   * this also means implementing user preferences elsewhere _should_ be a piece of cake
   * also note that if for whatever reason the GET request to `/preferences` fails, we are not going to `PUT /preferences` until a GET request succeeds. We ABSOLUTELY DO NOT want to overwrite a user's preferences if we can't first successfully know which ones are already set.
2. When you toggle a preference, the UI updates immediately, but any API calls are debounced for 500ms, to prevent too many network requests if the user keeps toggling back-and-forth
3. Each theme is now a function that takes options as an argument.
    * e.g. `dark({ spacingUnit: '4px' })` or `light({ spacingUnit: '8px' })`
4. All Redux Profile logic has been updated to be async instead of sync
    * in other words, we implemented thunks for getting and updating the profile.

## To Test

1. Try toggling back and forth between the theme and spacing
2. Try doing it very quickly
3. Try forcing errors or `GET /preferences` and `PUT /preferences`
4. Try toggling some more
5. Load the app while blocking `/preferences`
6. etc....

### CODE REVIEW

File: src/components/PreferenceToggle/PreferenceToggle.tsx [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+  React.memo(component, (prevProps, nextProps) => {
+    /**
+     * only prevent rendering if the preferences AND profileError
+     * went from being defined to defined or vise versa.
+     *
+     * we don't care to re-render if the preferences have been updated in Redux
+     * state. All the relevant preference state will be handled in the component.
+     * This component only cares about what the preferences are on app load.
+     */
+    // console.log(wasUndefinedNowDefined(prevProps.preferenceError, nextProps.preferenceError))
```
Comment (martinmckenna) [inline src/components/PreferenceToggle/PreferenceToggle.tsx]: remove this dummy
